### ISSUE TITLE
Lets fix MinIO together !

### ISSUE THREAD

**Issue by kgnfth** [caprover/one-click-apps#479]
Hello, i need your help fixing MinIO app

The latest release of MinIO `RELEASE.2021-08-05T22-01-19Z` will not work with the current [MinIO one-click-app](https://github.com/caprover/one-click-apps/blob/master/public/v4/apps/minio.yml)

Because the api listens on port 9000 and the dashboard on 9001.

So if we enable SSL and set `containerHttpPort` to 9001 we would have `HTTPS://minio.redacted.com`,
and if we open port 9000 we could use the api `HTTP://minio.redacted.com:9000` as our s3 endpoint!
but the problem is that this is not SSL enabled.. Wish we could enable SSL on port 9000 or something

For now i have deployed using this
```yaml
captainVersion: 4
services:
    $$cap_appname:
        volumes:
            - $$cap_appname-db-data:/data
            - $$cap_appname-config-data:/root/.minio
        restart: always
        environment:
            MINIO_ROOT_USER: $$cap_access_key
            MINIO_ROOT_PASSWORD: $$cap_secret_key
            MINIO_REGION_NAME: "eu-east-1"
            MINIO_BROWSER_REDIRECT_URL: https://$$cap_appname.$$cap_root_domain
            MINIO_SERVER_URL: http://$$cap_appname.$$cap_root_domain:9000
        ports:
            - 9000:9000
        caproverExtra:
            containerHttpPort: '9001'
            dockerfileLines:
                - FROM minio/minio:$$cap_minio_version
                - CMD ["server", "/data", "--console-address", ":9001"] #We can also change the api port 9000 with --address flag
caproverOneClickApp:
    variables:
        - id: $$cap_minio_version
          label: MinIO Version
          defaultValue: RELEASE.2021-08-05T22-01-19Z
          description: Check out their Docker page for the valid tags https://hub.docker.com/r/minio/minio/tags/
          validRegex: /^([^\s^\/])+$/
        - id: $$cap_access_key
          label: MinIO Access Key
          defaultValue: ''
          description: Username to access MinIO server
          validRegex: /.{5,}/
        - id: $$cap_secret_key
          label: Minio Secret Key
          defaultValue: ''
          description: Password to access MinIO server
          validRegex: /.{8,}/
    instructions:
        start: >-
            MinIO is a High Performance Object Storage released under GNU Affero General Public License v3.0. It is API compatible with Amazon S3 cloud storage service.
        end: Minio is deployed and available as $$cap_appname
    displayName: ''
    isOfficial: true
    description: MinIO is a cloud storage server compatible with Amazon S3
    documentation: Taken from https://docs.min.io/docs/minio-docker-quickstart-guide.html
```
With this we can access the dashboard at `HTTPS://minio.redacted.com` and s3 endpoint at `HTTP://minio.redacted.com:9000`

Without Caprover i can use my setup with `nginx` & `goacme/lego` using my repo https://github.com/kgnfth/minio-docker-certbot

With my [repo ](https://github.com/kgnfth/minio-docker-certbot) i can access api using SSL  `https://s3.redacted.com`

**Comment by Handfish** [comment]
This is a good start. With the provided template, the console webpage renders but I am unable to authenticate with either the provided environment variables or the `minioadmin:minioadmin` defaults.

**Comment by Handfish** [comment]
Using the `RELEASE.2021-07-27T02-40-15Z` tag, we get a more expressive console.
```
2021-08-17T17:03:23.183407384Z original error -> (user_login.go:50: invalid Login)
2021-08-17T17:03:23.183426680Z debugging error: Post "https://minio-test.{MY_URL}.9000/": dial tcp {MY_IP}:9000: connect: connection refused
```

**Comment by kgnfth** [comment]
> This is a good start. With the provided template, the console webpage renders but I am unable to authenticate with either the provided environment variables or the `minioadmin:minioadmin` defaults.
> 
> The interesting thing is the login form is getting valid responses from the application, but its always a 401.
> `{code: 401, detailedMessage: "invalid Login", message: "invalid Login"}`
> 
> I think it's the same issue happening here: [minio/console#871](https://github.com/minio/console/issues/871)

make sure that MINIO_SERVER_URL point to http://minio-test.{MY_URL}:9000

**Comment by Handfish** [comment]
> make sure that MINIO_SERVER_URL point to [http://minio-test.{MY_URL}:9000](http://minio-test.%7BMY_URL%7D:9000)

Yep, that is the value I have for that environment variable.

**Comment by kgnfth** [comment]
try `RELEASE.2021-08-05T22-01-19Z` instead of `RELEASE.2021-07-27T02-40-15Z`

Sorry for the late reply

**Comment by kgnfth** [comment]
> > make sure that MINIO_SERVER_URL point to [http://minio-test.{MY_URL}:9000](http://minio-test.%7BMY_URL%7D:9000)
> 
> Yep, that is the value I have for that environment variable.

Make sure port 9000 is not blocked by your firewall

Sorry for the late reply

**Comment by kgnfth** [comment]
> > make sure that MINIO_SERVER_URL point to [http://minio-test.{MY_URL}:9000](http://minio-test.%7BMY_URL%7D:9000)
> 
> Yep, that is the value I have for that environment variable.

the part here` 2021-08-17T17:03:23.183426680Z debugging error: Post "https://minio-test.{MY_URL}:9000/"` tries to connect to HTTPS instead of HTTP

Also make sure port 9000 is not blocked by your firewall

Sorry for the late reply

**Comment by githubsaturn** [comment]
@kgnfth  - that's definitely doable. We have a special nginx custom image that does this. Here is [an example of it in one click app](https://github.com/caprover/one-click-apps/blob/48d2870e6e5769acce66a4a14712968f347683bf/public/v4/apps/mailtrain.yml).


This is what you need to do:
- add a new service for this nginx reverse proxy
- set the image to `nginx-reverse-proxy:latest`
- set the env var to  `UPSTREAM_HTTP_ADDRESS: http://srv-captain--$$cap_appname:9000`
- also, remove the mapped port for 9000, it's no longer needed.

Now if you enable SSL for the main app, you'll get the dashboard covered, and if you enable SSL for the nginx app, you'll get the S3 endpoint covered.

**Comment by kgnfth** [comment]
> @kgnfth - that's definitely doable. We have a special nginx custom image that does this. Here is [an example of it in one click app](https://github.com/caprover/one-click-apps/blob/48d2870e6e5769acce66a4a14712968f347683bf/public/v4/apps/mailtrain.yml).
> 
> This is what you need to do:
> 
> * add a new service for this nginx reverse proxy
> * set the image to `nginx-reverse-proxy:latest`
> * set the env var to  `UPSTREAM_HTTP_ADDRESS: http://srv-captain--$$cap_appname:9000`
> * also, remove the mapped port for 9000, it's no longer needed.
> 
> Now if you enable SSL for the main app, you'll get the dashboard covered, and if you enable SSL for the nginx app, you'll get the S3 endpoint covered.

Thx it works but upon authentication we get the following error

```
2021-08-18T14:06:43.511381920Z API: SYSTEM()
2021-08-18T14:06:43.511384314Z Time: 14:06:43 UTC 08/18/2021
2021-08-18T14:06:43.511386198Z DeploymentID: 2050a0ba-7d92-4c83-8f14-9107e7841435
2021-08-18T14:06:43.511388051Z Error: The request signature we calculated does not match the signature you provided. Check your key and signing method. (*errors.errorString)
2021-08-18T14:06:43.511391107Z requestHeaders={"method":"GET","reqURI":"/minio/admin/v3/info","header":{"Authorization":["AWS4-HMAC-SHA256 Credential=deejavu/20210818//s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=b45afd173f101f67b15ca757f49db2f6364e7dd2e6f9faa7a93a66df18991c8c"],"Connection":["close"],"Host":["srv-captain--minio:9000"],"User-Agent":["MinIO (linux; amd64) madmin-go/0.0.1 mc/RELEASE.2021-07-27T06-46-19Z"],"X-Amz-Content-Sha256":["e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"],"X-Amz-Date":["20210818T140643Z"],"X-Forwarded-For":["172.18.0.1, 10.0.1.32"],"X-Forwarded-Proto":["http"],"X-Real-Ip":["10.0.1.32"]}}
2021-08-18T14:06:43.511396527Z 4: cmd/auth-handler.go:154:cmd.validateAdminSignature()
2021-08-18T14:06:43.511398421Z 3: cmd/auth-handler.go:165:cmd.checkAdminRequestAuth()
2021-08-18T14:06:43.511400224Z 2: cmd/admin-handlers.go:1891:cmd.adminAPIHandlers.ServerInfoHandler()
2021-08-18T14:06:43.511402158Z 1: net/http/server.go:2049:http.HandlerFunc.ServeHTTP()
```

This is the yaml i made

```yml
captainVersion: 4
services:
    # Minio Dashboard
    $$cap_appname: # = minio.domain.com
        volumes:
            - $$cap_appname-db-data:/data
            - $$cap_appname-config-data:/root/.minio
        restart: always
        environment:
            MINIO_ROOT_USER: $$cap_access_key
            MINIO_ROOT_PASSWORD: $$cap_secret_key
            MINIO_REGION_NAME: "eu-east-1"
            MINIO_BROWSER_REDIRECT_URL: https://$$cap_appname.$$cap_root_domain
            MINIO_SERVER_URL: https://$$cap_endointname.$$cap_root_domain     # MinIO S3 API
        caproverExtra:
            containerHttpPort: '9001'
            dockerfileLines:
                - FROM minio/minio:RELEASE.2021-08-17T20-53-08Z
                - CMD ["server", "/data", "--console-address", ":9001"]
    # MinIO S3 API
    $$cap_endointname: # = s3.domain.com
        image: caprover/nginx-reverse-proxy:latest
        restart: always
        depends_on:
            - $$cap_appname
        environment:
            UPSTREAM_HTTP_ADDRESS: http://srv-captain--$$cap_appname:9000
caproverOneClickApp:
    variables:
        - id: $$cap_minio_version
          label: MinIO Version
          defaultValue: RELEASE.2021-08-05T22-01-19Z
          description: Check out their Docker page for the valid tags https://hub.docker.com/r/minio/minio/tags/
          validRegex: /^([^\s^\/])+$/
        - id: $$cap_endointname
          label: "MinIO Endpoit name eg: s3"
          defaultValue: s3
          description: 'This will become appname.s3.yourdomain.com'
        - id: $$cap_access_key
          label: MinIO Access Key
          defaultValue: ''
          description: Username to access MinIO server
          validRegex: /.{5,}/
        - id: $$cap_secret_key
          label: Minio Secret Key
          defaultValue: ''
          description: Password to access MinIO server
          validRegex: /.{8,}/
    instructions:
        start: >-
            MinIO is a High Performance Object Storage released under GNU Affero General Public License v3.0. It is API compatible with Amazon S3 cloud storage service.
        end: Minio is deployed and available as $$cap_appname
    displayName: ''
    isOfficial: true
    description: MinIO is a cloud storage server compatible with Amazon S3
    documentation: Taken from https://docs.min.io/docs/minio-docker-quickstart-guide.html
```

**Comment by githubsaturn** [comment]
Looks like there are some more headers that need to be added to the nginx proxy, see here:
- https://github.com/minio/minio/issues/7936
- https://github.com/minio/minio/issues/8999


I've update the nginx and pushed a new image under this tag: `1-ef5ffcb`. So change your 
```
caprover/nginx-reverse-proxy:latest
```
to 
```
caprover/nginx-reverse-proxy:1-ef5ffcb
```

and try again.

**Comment by Handfish** [comment]
@githubsaturn 
Your changes nginx proxy changes finally made the web-console login work properly, and the rest of minio as well. Everything looks good, I'm able to interact with it through the cli tool and create / delete buckets.

My deployment has SSL handled by a load balancer in front of Caprover, so my results can't speak for everyone.

Thanks a lot to @kgnfth and @githubsaturn for their work!

**Comment by kgnfth** [comment]
@githubsaturn YaY it works now!

**Comment by githubsaturn** [comment]
Great! Feel free to open a PR.

**Comment by italodeandra** [comment]
Hi @kgnfth. Do you plan on contributing the update as a PR? That would be really nice ❤️

**Comment by githubsaturn** [comment]
@italodeandra - I've just updated the definition. Please test and let me know if everything works.

**Comment by kgnfth** [comment]
> @italodeandra - I've just updated the definition. Please test and let me know if everything works as I'm not a Minio user and I don't know what/how to test.
> 
> [fb8ad02](https://github.com/caprover/one-click-apps/commit/fb8ad02a4dc49145c226757a39f52875303a5fa0)

Thanks for updating, i could not update because i have a hard time i dont want to go into the detail...
Anyway the code looks fine and should work, only downside is that we need to update the image manually everytime there is an [release](https://github.com/minio/minio/releases)

eg: whe have to run 
```
FROM minio/minio:RELEASE.2021-09-03T03-56-13Z
CMD ["server", "/data", "--console-address", ":9001"]
```

using method 4 from the deployments page

**Comment by kgnfth** [comment]
> @italodeandra - I've just updated the definition. Please test and let me know if everything works as I'm not a Minio user and I don't know what/how to test.
> 
> [fb8ad02](https://github.com/caprover/one-click-apps/commit/fb8ad02a4dc49145c226757a39f52875303a5fa0)

Thanks for updating, i could not update because i have a hard time i dont want to go into the detail...
Anyway the code looks fine and should work, only downside is that we need to update the image manually everytime there is an [release](https://github.com/minio/minio/releases)

eg: whe have to run 
```
FROM minio/minio:RELEASE.2021-09-03T03-56-13Z
CMD ["server", "/data", "--console-address", ":9001"]
```

using method 4 from the deployments page

![image](https://user-images.githubusercontent.com/73314940/132282206-7d2c1103-ebd5-463d-8920-c455b1f4b3db.png)

**Comment by italodeandra** [comment]
@githubsaturn I created a new instance with the newest version and it's working like a charm. Thanks!

**Comment by uhlhosting** [comment]
@githubsaturn @kgnfth unfortunately updating Minio to latest versions fails in all possible ways. 

At first it does not uploads the image anymore in the local registry, and second it remains on a 502 classic error page, I believe that each one-click-app should have a clear described upgrade path. 

This is not a game of guessing, some apps works to be updated by their docker tags/name:version while attempted to do same with Minio will simply not work.
